_posts/tutorials/2018-03-01-javalin-kotlin-integration-testing-example.md

```diff
@@ -0,0 +1,254 @@
+---
+layout: tutorial
+title: "Integration Testing with khttp"
+author: <a href="https://www.linkedin.com/in/kevin-cianfarini-b37807101/" target="_blank">Kevin Cianfarini</a>
+date: 2018-03-01
+permalink: /tutorials/integration-testing-kotlin
+github: https://github.com/kevincianfarini/fastfood
+summarytitle: Integration Testing With khttp
+summary: Learn How to Build Integration Tests For Your Server
+language: kotlin
+---
+
+## What You'll Create
+In this tutorial, you'll be walked through creating a very limited CRUD app and testing it with JUnit, Khttp, and Javalin. 
+
+
+## Dependencies 
+First, get your workspace set up with gradle. Then you're going to need to add some dependencies to your project. Below are the relevant bits that you should pay attention to. 
+
+```
+repositories {
+    mavenCentral()
+    jcenter()
+}
+
+dependencies {
+    compile 'org.jetbrains.kotlin:kotlin-stdlib-jdk8:$kotlin_version
+    compile 'io.javalin:javalin:1.3.0'
+    compile "org.slf4j:slf4j-api:1.7.25"
+    compile "org.slf4j:slf4j-simple:1.7.25"
+    compile group: 'com.fasterxml.jackson.core', name: 'jackson-databind', version: '2.9.4'
+    compile 'com.google.code.gson:gson:2.8.2'
+    testCompile 'junit:junit:4.12'
+    testCompile 'khttp:khttp:0.1.0'
+}
+```
+
+The library we're going to be using for testing is `khttp`, an extremely lightweight http implementatiton in kotlin. If you've ever used pythons `requests` module, you'll notice a lot of similarities. 
+
+## Getting Started
+
+I set up my file structure in an MVC-like fashion as follows. 
+
+```
+src
+├── main
+│   └── kotlin
+│       ├── controllers
+│       │   ├── FoodItemController.kt
+│       │   └── TestController.kt
+│       ├── Main.kt
+│       ├── models
+│       │   └── FoodItem.kt
+│       └── util
+│           └── Extensions.kt
+└── test
+    └── kotlin
+        └── integration
+            └── TestIntegration.kt
+```
+
+
+First and foremost, to be able to test anything we need to be able to start and stop the server from our tests. The way the docs outline how to do this in a functional manner makes it difficult to do this. 
+```
+// Main.kt
+fun main(args: Array<String>) {
+    val app = Javalin.start(7000)
+    app.get("/") { ctx -> ctx.result("Hello World") }
+}
+```
+A simple workaround for this is wrapping all of that in a dummy class. 
+
+```
+// Main.kt
+fun main(args: Array<String>) {
+
+    JavalinApp(8000).create()
+
+}
+
+
+class JavalinApp(private val port: Int) {
+
+    fun create(): Javalin {
```

Comment:
Perhaps call this `init()` or `start()` to avoid confusion? `Javalin.create()` just creates a server.